File: solutions/python/King's_order.py
import heapq
from sys import stdin, stdout

class Project:
    def __init__(self, id, group):
        self.id = id
        self.group = group
    
    def __lt__(self, other):
        if self.group == other.group:
            return self.id < other.id
        return self.group < other.group

n, m = map(int, input().split())
group_id = list(map(int, input().split()))
adj = [[] for _ in range(n)]
in_degree = [0] * n

for _ in range(m):
    a, b = map(int, input().split())
    a -= 1
    b -= 1
    adj[a].append(b)
    in_degree[b] += 1

pq = []
for i in range(n):
    if in_degree[i] == 0:
        heapq.heappush(pq, Project(i, group_id[i]))

result = []

while pq:
    current = heapq.heappop(pq)
    result.append(current.id + 1)
    
    for neighbor in adj[current.id]:
        in_degree[neighbor] -= 1
        if in_degree[neighbor] == 0:
            heapq.heappush(pq, Project(neighbor, group_id[neighbor]))

if len(result) != n:
    print(-1)
else:
    print(" ".join(map(str, result)))

File: solutions/cpp/King's_order.cpp
#include <bits/stdc++.h>
using namespace std;

struct Project {
    int id;
    int group;
    Project(int id, int group) : id(id), group(group) {}
    bool operator<(const Project &other) const {
        if (group == other.group) return id > other.id;
        return group > other.group;
    }
};

int main() {
    int n, m;
    cin >> n >> m;
    vector<int> group_id(n);
    vector<vector<int>> adj(n);
    vector<int> in_degree(n, 0);
    
    for (int i = 0; i < n; ++i) {
        cin >> group_id[i];
    }

    for (int i = 0; i < m; ++i) {
        int a, b;
        cin >> a >> b;
        --a; --b;
        adj[a].push_back(b);
        in_degree[b]++;
    }

    priority_queue<Project> pq;
    
    for (int i = 0; i < n; ++i) {
        if (in_degree[i] == 0) {
            pq.emplace(i, group_id[i]);
        }
    }

    vector<int> result;

    while (!pq.empty()) {
        auto [current, group] = pq.top(); pq.pop();
        result.push_back(current + 1);

        for (int neighbor : adj[current]) {
            if (--in_degree[neighbor] == 0) {
                pq.emplace(neighbor, group_id[neighbor]);
            }
        }
    }

    if (result.size() != n) {
        cout << -1 << endl;
    } else {
        for (int proj : result) {
            cout << proj << " ";
        }
        cout << endl;
    }

    return 0;
}
